daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala

```diff
@@ -167,8 +184,15 @@ private[lf] final class CommandPreprocessor(
     cmd match {
       case command.ReplayCommand.Create(templateId, argument) =>
         unsafePreprocessCreate(templateId, argument)
-      case command.ReplayCommand.Exercise(templateId, coid, choiceId, argument) =>
-        unsafePreprocessExercise(templateId, coid, choiceId, argument)
+      case command.ReplayCommand.LenientExercise(typeId, coid, choiceId, argument) =>
+        unsafePreprocessLenientExercise(typeId, coid, choiceId, argument)
+      case command.ReplayCommand.Exercise(templateId, mbIfaceId, coid, choiceId, argument) =>
+        mbIfaceId match {
+          case Some(ifaceId) =>
+            unsafePreprocessExerciseInterface(ifaceId, coid, choiceId, argument)
```

Comment:
If you have both an interfaceId and templateId, you should pass both to the preprocessor, and use the templateid in the exercise guard (via the speedy command `ExerciseByInterface`)

Reply:
There's also an argument to be made here that you should be able to pass the requiring interface id as the "template id" and the required interface id as the "interface id", and have that interpreted by the `ExerciseByInheritedInterface.`

Reply:
Oh if this is replay commands then this last point shouldn't matter, since you know the actually know the template id. I thought this was api commands.

Reply:
My PR drops the possibility to do type checking in the ledger API as you described here.
With my change, you have:

In case of template choice you use the templateID
in case of interface choice you use the interfaceID.

I wanted my change on the API to be minimal. The PR just adds the optional interfaceId filed to the ExerciseEvent API output. No change to input command (in particular we still have unique identifier in the ExerciceCommand).

With my changes, one will need to use an intermediary CreateAndExercise to perform the type checking (or any other guard).

I am fine to add the field in the Exercise Command in an upcoming PR.

Reply:
I was confused -- I thought I was looking at changes to API command, but this is a change in the Replay command. Here for replay commands, the template id argument always refers to a template id, and the interface id always refers to the interface that defines the choice (if the choice was defined in an interface). And since it's a replay we don't need the extra type checks in the guard. So this looks fine. 👍